### ISSUE TITLE
Cypress hangs on empty tests files

### ISSUE THREAD

**Issue by paulfalgout** [cypress-io/cypress#1905]
### Current behavior:

In v3 we have a few test files that don't yet contain tests.  In v2 it wasn't a problem, and most of the time it's fine in v3.  But occasionally Cypress will hang on these files.  I've only seen this in CI so far.
<img width="297" alt="screen shot 2018-06-08 at 3 01 58 pm" src="https://user-images.githubusercontent.com/2028470/41141543-3d316efa-6b2d-11e8-9d73-3078698ca68e.png">

In the dashboard these runs still say "running" with an increasing duration, even though they're not.

Further these panels aren't super useful as due to our folder structure every panel is labeled "assets/tests/integrati..."  I can resize my screen to see a little more, but never the file name.

Otherwise really loving the v3 direction!

### Desired behavior:

Shouldn't hang and should produce better errors if they do.

### Steps to reproduce:

Unsure other than empty test files and multiple runs

### Versions

Cypress 3.0.1

**Comment by paulfalgout** [comment]
Related to the "running" status part of this: https://github.com/cypress-io/cypress/issues/1891  Looks like that may be fixed

**Comment by jennifer-shehane** [comment]
We're updating this entire view in the Dashboard pretty soon. We've been using it and also have not been satisfied with various things like the spec name being cut off!

I created a repo with an empty test file and am unable to recreate an 'error' state. The tests show up as status 'No Tests':

![screen shot 2018-06-08 at 11 39 07 am](https://user-images.githubusercontent.com/1271364/41167363-a47a1c3c-6b10-11e8-8061-3a2694750c94.png)

- Do you have any code within the file?
- Do you have any `before`, `after` hooks, or code in your `support` files?
- Do you have any add-on code set up in your `plugins` file?
- What CI are you running from? When clicking on "See CI logs", what are those logs?

**Comment by paulfalgout** [comment]
We didn't have any code in the file.. however now i'm seeing it on files with code as well.
There are no hooks in our support files, but we have many `Cypress.Commands.add`s
We're using `@cypress/webpack-preprocessor` in plugins but that's it.
We're running travis
```                                                                                              
  Running: clinicians/clinician-profile.js...                                             (2 of 43) 
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated
```
The problem is inconsistent, but when it does fail it seems to be in the same spots.  It always seem to fail prior to any tests starting in a new file.
I can't seem to get it to happen locally.  Seems like this might be a duplicate of https://github.com/cypress-io/cypress/issues/1912 and https://github.com/cypress-io/cypress/issues/1890 We weren't experiencing this in v2. 

There's nothing else really in the log currently.  Anyway to force more output for travis?

**Comment by paulfalgout** [comment]
Definitely this one: https://github.com/cypress-io/cypress/issues/1934

**Comment by jennifer-shehane** [comment]
The only logging information can be printed by setting `DEBUG=cypress:*` env var in `cypress run`

**Comment by paulfalgout** [comment]
welp if I turn on debug in travis I get `The log length has exceeded the limit of 4 MB (this usually means that the test suite is raising the same exception over and over).`  But I don't think it reaches any error.. just a log of log dump

**Comment by paulfalgout** [comment]
```

**Comment by paulfalgout** [comment]
not sure what to log.. can't log everything and travis can't handle it.. nothing is displayed and not tests start for the file it crashes on.. doesn't happen on any particular file consistently.

**Comment by brian-mann** [comment]
@paulfalgout I don't think you need to do anything. We've collected data from about 4 other situations that are experiencing the same bug.

We're working on it now. We believe it has to do with a file lock never releasing, which causes a Promise to hang that never resolves.

Can you try to run your tests in Chrome? We believe this problem only has to do with Electron, but don't quote me on that just yet.

**Comment by brian-mann** [comment]
Most definitely the same issue as this: https://github.com/cypress-io/cypress/issues/1912 and https://github.com/withspectrum/spectrum/pull/3334#issuecomment-397671374 which failed in Circle job: https://circleci.com/gh/withspectrum/spectrum/8642

We also have reports from two other organizations with the same problem.

**Comment by paulfalgout** [comment]
👍 ok trying chrome

**Comment by paulfalgout** [comment]
@brian-mann yep so this time a couple of times it said `Timed out waiting for the browser to connect. Retrying...` but it finished 🎉  It also seems to be faster.   I think at least for me for now, this is resolved.

**Comment by paulfalgout** [comment]
v3 never fails with the browser set to chrome, but runs are averaging between 15-18 mins.  In v2 we were getting between 5-7 mins.  For the moment we're reverting.

**Comment by brian-mann** [comment]
We have noticed that Chrome does indeed run faster. The v3 performance problems are a separate issue altogether that we are looking at. We believe we have the fixes for the spec file just hanging as per this PR: https://github.com/cypress-io/cypress/pull/2020

**Comment by paulfalgout** [comment]
Good deal thanks.  I'll follow that PR and re-upgrade when it's released

**Comment by brian-mann** [comment]
Released in [`3.0.2`](https://docs.cypress.io/guides/references/changelog.html#3-0-2).

**Comment by kamituel** [comment]
I'm sorry to be the bearer of bad news, but I think this remains an issue on `3.0.2` (I didn't take part in this discussion, but we've been experiencing it since at least `3.0.1` - not sure about `3.0.0`).

Our build **hang 4 times** out of 10 runs today.

Here are some clues I've noticed (or might be useful):

1. When running with `DEBUG=cypress:*`, test runs hang less frequently.

2. From all the times I've managed to capture logs of it hanging, the last meaningful log message always is:

> cypress:server:file reading JSON file /root/.config/Cypress/cy/production/projects/project-name-71948c40449279172805d15f83be086b/state.json

All the subsequent log statements are like this:

> cypress:server:timers child sending timer id 903
> cypress:server:timers child sending timer id 920
> cypress:server:timers child sending timer id 921
> cypress:server:timers child sending timer id 922
> cypress:server:timers child sending timer id 924

3. This led me to `open()` in `server/lib/browsers/electron.coffee`. [There's a promise there](https://github.com/cypress-io/cypress/blob/develop/packages/server/lib/browsers/electron.coffee#L125) that either resolves immediately with a concrete value of the `options` map, or invokes the `before:browser:launch` and returns a promise. We've not been using this hook, so I added a dummy one. This seems to decrease the likelihood of test run hanging, but doesn't fix it completely (got 1 hang out of 10 test runs today, compared to 4 hangs without this plugin - but the sample size is small, so it might've been a chance).

4. Our Cypress tests are a bit unorthodox in that some spec files are empty (have no `describe()`'s). That's a long story, and I don't have any reason to believe that's the source of the issue - mentioning this just in case it's relevant.

We're using Bamboo, with Cypress running within a Docker container (custom image, not based on images provided by Cypress).

We can't upgrade to Cypress `3.0.x` because of that :/ so I'll be grateful for any tips. I'll be also happy to help debug this, if needed.

**Comment by kamituel** [comment]
Oh, I actually managed to capture logs on `3.0.2` just now. 

Here's the log for the suite that hanged:

```
  Running: <obscured>.js...                                                      (17 of 18)
Mon, 02 Jul 2018 12:54:23 GMT cypress:server:run about to run spec { spec: { name: '<obscured>.js', relative: 'cypress/integration/<obscured>.js', absolute: '/<obscured>/cypress/integration/<obscured>.js' }, isHeadless: true, browserName: 'electron' }
Mon, 02 Jul 2018 12:54:23 GMT cypress:server:video capture started { command: 'ffmpeg -n 20 /root/.cache/Cypress/3.0.2/Cypress/resources/app/packages/server/node_modules/@ffmpeg-installer/linux-x64/ffmpeg -f image2pipe -use_wallclock_as_timestamps 1 -i pipe:0 -y -vcodec libx264 -preset ultrafast /results/videos/<obscured>js.mp4' }
Mon, 02 Jul 2018 12:54:23 GMT cypress:server:openproject resetting project state, preparing to launch browser
Mon, 02 Jul 2018 12:54:23 GMT cypress:server:project resetting project instance <obscured>
Mon, 02 Jul 2018 12:54:23 GMT cypress:server:server Setting remoteAuth undefined
Mon, 02 Jul 2018 12:54:23 GMT cypress:server:server Setting remoteOrigin http://localhost:44476
Mon, 02 Jul 2018 12:54:23 GMT cypress:server:server Setting remoteStrategy file
Mon, 02 Jul 2018 12:54:23 GMT cypress:server:server Setting remoteHostAndPort null
Mon, 02 Jul 2018 12:54:23 GMT cypress:server:server Setting remoteDocDomain localhost
Mon, 02 Jul 2018 12:54:23 GMT cypress:server:server Setting remoteFileServer http://localhost:48398
Mon, 02 Jul 2018 12:54:23 GMT cypress:server:server Getting remote state: { auth: undefined, props: null, origin: 'http://localhost:44476', strategy: 'file', visiting: false, domainName: 'localhost', fileServer: 'http://localhost:48398' }
Mon, 02 Jul 2018 12:54:23 GMT cypress:server:timers queuing timer id 273 after 30000 ms
Mon, 02 Jul 2018 12:54:23 GMT cypress:server:openproject launching browser: electron, spec: cypress/integration/<obscured>.js
Mon, 02 Jul 2018 12:54:23 GMT cypress:server:timers child received timer id 273
Mon, 02 Jul 2018 12:54:23 GMT cypress:server:browsers opening browser electron
Mon, 02 Jul 2018 12:54:23 GMT cypress:server:saved_state noop saved state
Mon, 02 Jul 2018 12:54:23 GMT cypress:server:video capture stderr log { message: 'ffmpeg version N-45896-g19c3df0cd-static https://johnvansickle.com/ffmpeg/  Copyright (c) 2000-2018 the FFmpeg developers' }
Mon, 02 Jul 2018 12:54:23 GMT cypress:server:video capture stderr log { message: '  built with gcc 6.3.0 (Debian 6.3.0-18+deb9u1) 20170516' }
Mon, 02 Jul 2018 12:54:23 GMT cypress:server:video capture stderr log { message: '  configuration: --enable-gpl --enable-version3 --enable-static --disable-debug --disable-ffplay --disable-indev=sndio --disable-outdev=sndio --cc=gcc-6 --enable-fontconfig --enable-frei0r --enable-gnutls --enable-gray --enable-libaom --enable-libfribidi --enable-libass --enable-libvmaf --enable-libfreetype --enable-libmp3lame --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-librubberband --enable-libsoxr --enable-libspeex --enable-libvorbis --enable-libopus --enable-libtheora --enable-libvidstab --enable-libvo-amrwbenc --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libxvid --enable-libzimg' }
Mon, 02 Jul 2018 12:54:23 GMT cypress:server:video capture stderr log { message: '  libavutil      56. 17.100 / 56. 17.100' }
Mon, 02 Jul 2018 12:54:23 GMT cypress:server:video capture stderr log { message: '  libavcodec     58. 19.100 / 58. 19.100' }
Mon, 02 Jul 2018 12:54:23 GMT cypress:server:video capture stderr log { message: '  libavformat    58. 13.100 / 58. 13.100' }
Mon, 02 Jul 2018 12:54:23 GMT cypress:server:video capture stderr log { message: '  libavdevice    58.  4.100 / 58.  4.100' }
Mon, 02 Jul 2018 12:54:23 GMT cypress:server:video capture stderr log { message: '  libavfilter     7. 20.100 /  7. 20.100' }
Mon, 02 Jul 2018 12:54:23 GMT cypress:server:video capture stderr log { message: '  libswscale      5.  2.100 /  5.  2.100' }
Mon, 02 Jul 2018 12:54:23 GMT cypress:server:video capture stderr log { message: '  libswresample   3.  2.100 /  3.  2.100' }
Mon, 02 Jul 2018 12:54:23 GMT cypress:server:video capture stderr log { message: '  libpostproc    55.  2.100 / 55.  2.100' }
Mon, 02 Jul 2018 12:54:24 GMT cypress:server:timers child sending timer id 166
Mon, 02 Jul 2018 12:54:26 GMT cypress:server:timers child sending timer id 179
Mon, 02 Jul 2018 12:54:27 GMT cypress:server:timers child sending timer id 180
Mon, 02 Jul 2018 12:54:30 GMT cypress:server:timers child sending timer id 193
Mon, 02 Jul 2018 12:54:31 GMT cypress:server:timers child sending timer id 194
Mon, 02 Jul 2018 12:54:33 GMT cypress:server:timers child sending timer id 207
Mon, 02 Jul 2018 12:54:34 GMT cypress:server:timers child sending timer id 208
Mon, 02 Jul 2018 12:54:36 GMT cypress:server:timers child sending timer id 221
Mon, 02 Jul 2018 12:54:37 GMT cypress:server:timers child sending timer id 222
Mon, 02 Jul 2018 12:54:40 GMT cypress:server:timers child sending timer id 240
Mon, 02 Jul 2018 12:54:40 GMT cypress:server:timers child sending timer id 241
Mon, 02 Jul 2018 12:54:40 GMT cypress:server:timers child sending timer id 242
... (many more logs from cypress:sever:timers) ...
```

**Comment by kamituel** [comment]
@brian-mann @jennifer-shehane I see this issue has been removed from `3.0.2` milestone and closed again. What does it mean? Last time I heard, there were plans to add more verbose logging and let us run this to reproduce. Is it still the case? Is it going to happen on `3.0.2`?

**Comment by jennifer-shehane** [comment]
The fix did not make it into 3.0.2 and is instead included in version 3.0.3 - which is pending release. We close issues that have a closed PR associated with it, so be on the lookout for 3.0.3 official release here soon.

**Comment by paulfalgout** [comment]
So I've tried upgrading to `3.0.3`.  I'm running `--browser chrome` and I'm running `video: false`.

So far I haven't seen any outright failures, but I'm still getting quite a bit of:
```
Timed out waiting for the browser to connect. Retrying...
Timed out waiting for the browser to connect. Retrying again...
```
And when these happen the tests are way slower than they were in 2.x.  It's usually over double the run time, but I'm not sure why.

Any ideas?

**Comment by paulfalgout** [comment]
Found one thing.  If chrome stable isn't explicitly set on travis it uses chromium 62 which seems to perform _much_ worse than chrome 68. https://docs.travis-ci.com/user/chrome#selecting-a-chrome-version

It seems to be doing a little bit better.. though I'm not through the first run yet and it's still quite a bit slower.

**Comment by paulfalgout** [comment]
I think part of the reason it's slow is because it re-runs webpack every time.  We are using quite a few plugins and so our webpack isn't exactly quick.  Wasn't a problem in v2 I think because it packed at the beginning, but it's a lot of overhead to rerun it each time.

**Comment by paulfalgout** [comment]
No optimization I tried with the webpack preprocessor made any difference.  It also ran quite a bit slower on CI than locally.

I'm sure this defeats the purpose of the split in v3, and it certainly makes the dashboard less pretty but.. if I serialize all of the tests in one file like:
```js
describe('Integration Tests', function() {
    const req = require.context('./', true, /^(.*\.(js$))[^.]*$/i);
    req.keys().forEach(function(key) {
        req(key);
    });
});
```
... the tests run at the same rate, maybe faster, but the overall cypress time goes from 18 mins down to under 8 which is faster than it was with 2.1.0.  We're shooting for < 10min so I can move forward with this, but realistically it'd be ideal to get all of the tests preprocessed/built in on go and then split for the runner somehow..

**Comment by kamituel** [comment]
@brian-mann @jennifer-shehane I can confirm `3.1.0` indeed fixes that issue! While previously out of 10 CI runs, we've seen Cypress hang 4 times, yesterday it worked flawlessly every single time.

Kudos for fixing  that! We can finally upgrade from `2.x.x` :)
